### DISCUSSION TITLE
Exclude from playlist tracks from followed artists

### DISCUSSION THREAD

**Discussion by ffmusic** [Chimildic/goofy#90] [Помощь с алгоритмом] [answered]
Hi to you all!
I’m trying to exclude from a playlist the tracks from artists that I follow. 
I tried with something like:
Filter.rangeTracks(tracks, {artist: {followed_include: false,},}
but that doesn’t seem to work. 
Is there a way for doing it?
Thanks

**Comment by Chimildic** [comment, marked as answer]
You wrong when chosen [rangeTracks](https://chimildic.github.io/goofy/#/func?id=rangetracks). The function haven't param as `{artist: {followed_include: false,}`.

If you want get tracks from followed artists except some items, use the next code:
```js
let tracks = Source.getArtistsTracks({
    artist: {
        followed_include: true,
        exclude:  [
            { id: 'abc1', name: '' }, 
            { id: 'abc2', name: '' },
        ],
    },
});
```

If you have tracklist and want to remove tracks from followed artists, use the next code
```js
let artistTracks = Source.getArtistsTracks({ artist: { followed_include: true, }, });
let tracklist = Source.getPlaylistTracks('', 'id');
Filter.removeTracks(tracklist, artistTracks);
```

> Warning: function `getArtistsTracks` calls many requests in order to get all tracks from albums and singles. If you have many followed artists, runtime can be long. In addition, you spend a plenty part of [daily request limit](https://chimildic.github.io/goofy/#/desc?id=Ограничения).

Probably, we can do less expensive algorithm. Can you explain your idea for new playlist?

**Reply by ffmusic** [reply]
Thank you for your answer.
I follow a large playlist "A" (1000+ tracks) that is updated by the owner once a week with music from a genre I like.
Usually there are about 100 new songs each week and the same number of tracks is deleted by the author.
I want to discover new artists from that playlist.
With Goofy I make a copy of that playlist. Let's call it playlist "B".
When I really  like a song I save the track to "my liked songs" and I follow the artist.
If I don't like it, I add the track to a  playlist called "C".
Tracks from "liked songs" and playlist C are excluded from playlist B (easy to do it with Goofy).
I would also like to exclude other songs from artists that I already follow, as I'm only interested in discovering songs from artists that I don't know yet.
Could there be an easy way of doing this without spending the "daily request limit"?

**Reply by Chimildic** [reply]
In this case you needn't to _tracks_ from followed. Enough only artists. This fast and free way.

```js
let tracklist = // ...
let artists = Source.getArtists({ followed_include: true, });
Filter.removeArtists(tracklist, artists);
```

**Reply by ffmusic** [reply]
Thank you Chimildic!
I will try it.

**Reply by ffmusic** [reply]
It works exactly as I wanted.
BTW, thank you for your work (Goofy is a very well thought out project) .
I have some more doubts, I will ask for your help later.
Cheers,

FF

**Reply by Chimildic** [reply]
Recommend it to your friends)
